scipy/optimize/_brent.py

```diff
@@ -0,0 +1,181 @@
+    successive parabolic interpolation.
+
+    Let ```tol = rtol * abs(x0) + atol```, where ```x0``` is the best guess
+    found so far. It converges if evaluating a next guess would imply
+    evaluating ```f``` at a point that is closer than ```tol``` to a previously
+    evaluated one or if the number of iterations reaches ```maxiter```.
+
+    Args:
+        f (object): Objective function to be minimized.
+        a, b (float): endpoints of the interval a <= b.
+        rtol (float): relative tolerance. Defaults to 1.4901161193847656e-08.
+        atol (float): absolute tolerance. Defaults to 1.4901161193847656e-08.
+        maxiter (int): maximum number of iterations
+
+
+    Returns:
+        float: best guess for the minimum of f.
+        float: value of f evaluated at the best guess.
+        int: number of iterations performed.
+
+    References:
+        - http://people.sc.fsu.edu/~jburkardt/c_src/brent/brent.c
+        - Numerical Recipes 3rd Edition: The Art of Scientific Computing
+        - https://en.wikipedia.org/wiki/Brent%27s_method
+    """
+    # a, b: interval within the minimum should lie
+    #       no function evaluation will be requested
+    #       outside that range.
+    # x0: least function value found so far (or the most recent one in
+    #                                            case of a tie)
+    # x1: second least function value
+    # x2: previous value of x1
+    # (x0, x1, x2): Memory triple, updated at the end of each interation.
+    # u : point at which the function was evaluated most recently.
+    # m : midpoint between the current interval (a, b).
+    # d : step size and direction.
+    # e : memorizes the step size (and direction) taken two iterations ago
+    #      and it is used to (definitively) fall-back to golden-section steps
+    #      when its value is too small (indicating that the polynomial fitting
+    #      is not helping to speedup the convergence.)
+    #
+    #
+    # References: Numerical Recipes: The Art of Scientific Computing
+    # http://people.sc.fsu.edu/~jburkardt/c_src/brent/brent.c
+
+    assert a <= b
+    x0 = a + _golden * (b - a)
+    x1 = x0
+    x2 = x1
+    niters = -1
+    d = 0.0
+    e = 0.0
+    fx0 = f(x0)
+    fx1 = fx0
+    fx2 = fx1
+
+    for niters in range(maxiter):
+
+        m = 0.5 * (a + b)
+        tol = rtol * abs(x0) + atol
+        tol2 = 2.0 * tol
+
+        # Check the stopping criterion.
+        if abs(x0 - m) <= tol2 - 0.5 * (b - a):
+            break
+
+        r = 0.0
+        q = r
+        p = q
+
+        # "To be acceptable, the parabolic step must (i) fall within the
+        # bounding interval (a, b), and (ii) imply a movement from the best
+        # current value x0 that is less than half the movement of the step
+        # before last."
+        #   - Numerical Recipes 3rd Edition: The Art of Scientific Computing.
+
+        if tol < abs(e):
+            # Compute the polynomial of the least degree (Lagrange polynomial)
```

Comment:
Is this numerically stable enough? I thought there was a more stable way to go directly to the location of the minimum.